src/mlpack/methods/ann/layer/atrous_convolution_impl.hpp

```diff
@@ -0,0 +1,339 @@
+    const size_t outSize,
+    const size_t kW,
+    const size_t kH,
+    const size_t dW,
+    const size_t dH,
+    const size_t padW,
+    const size_t padH,
+    const size_t inputWidth,
+    const size_t inputHeight,
+    const size_t dilationW,
+    const size_t dilationH) :
+    inSize(inSize),
+    outSize(outSize),
+    kW(kW),
+    kH(kH),
+    dW(dW),
+    dH(dH),
+    padW(padW),
+    padH(padH),
+    inputWidth(inputWidth),
+    inputHeight(inputHeight),
+    outputWidth(0),
+    outputHeight(0),
+    dilationW(dilationW),
+    dilationH(dilationH)
+{
+  weights.set_size((outSize * inSize * kW * kH) + outSize, 1);
+}
+
+template<
+    typename ForwardConvolutionRule,
+    typename BackwardConvolutionRule,
+    typename GradientConvolutionRule,
+    typename InputDataType,
+    typename OutputDataType
+>
+void AtrousConvolution<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    InputDataType,
+    OutputDataType
+>::Reset()
+{
+    weight = arma::cube(weights.memptr(), kW, kH,
+        outSize * inSize, false, false);
+    bias = arma::mat(weights.memptr() + weight.n_elem,
+        outSize, 1, false, false);
+}
+
+template<
+    typename ForwardConvolutionRule,
+    typename BackwardConvolutionRule,
+    typename GradientConvolutionRule,
+    typename InputDataType,
+    typename OutputDataType
+>
+template<typename eT>
+void AtrousConvolution<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    InputDataType,
+    OutputDataType
+>::Forward(const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  inputTemp = arma::cube(input.memptr(), inputWidth, inputHeight, inSize);
+
+  if (padW != 0 || padH != 0)
+  {
+    Pad(inputTemp, padW, padH, inputPaddedTemp);
+  }
+
+  size_t wConv = ConvOutSize(inputWidth, kW, dW, padW, dilationW);
+  size_t hConv = ConvOutSize(inputHeight, kH, dH, padH, dilationH);
+
+  outputTemp = arma::zeros<arma::Cube<eT> >(wConv, hConv, outSize);
+
+  for (size_t outMap = 0, outMapIdx = 0; outMap < outSize; outMap++)
+  {
+    for (size_t inMap = 0; inMap < inSize; inMap++, outMapIdx++)
+    {
+      arma::Mat<eT> convOutput;
+      std::cout << "Fine Till Forward Convolution" << std::endl;
+      if (padW != 0 || padH != 0)
+      {
+        ForwardConvolutionRule::Convolution(inputPaddedTemp.slice(inMap),
+            weight.slice(outMapIdx), convOutput, dW, dH, dilationW, dilationH);
+      }
+      else
+      {
+        ForwardConvolutionRule::Convolution(inputTemp.slice(inMap),
+            weight.slice(outMapIdx), convOutput, dW, dH, dilationW, dilationH);
+      }
+      std::cout << "convOutput" << std::endl << convOutput << std::endl;
+      outputTemp.slice(outMap) += convOutput;
+    }
+
+    outputTemp.slice(outMap) += bias(outMap);
+  }
+
+  outputWidth = outputTemp.n_rows;
+  outputHeight = outputTemp.n_cols;
+
+  outputTemp.reshape(outputTemp.n_elem, 1, 1);
+  output = std::move(outputTemp.slice(0));
```

Comment:
If the idea of setting `outputTemp` to point to the same memory as `output` isn't feasible, I think this is a fine way to do it.